# **Part 1:**
we will calculate the marginal probabilities for each row and column and then check whether the product of the marginal probabilities for each cell equals the corresponding p_ij.

The calculations confirm that the observed probabilities match the expected probabilities under the null hypothesis of independence for each cell of the table. Specifically, the product of the marginal probabilities for each row and column is equal to the joint probability for each cell.

This indicates that the rows and columns in the given table are indeed independent, satisfying the null hypothesis. The table of observed probabilities matches the table of expected probabilities under the null hypothesis, and the boolean array confirms that all corresponding elements are equal, validating the hypothesis of independence.We fail to reject the null hypothesis.

In [1]:
import numpy as np

# Observed probabilities from the 3x3 table
observed_probabilities = np.array([
    [0.15, 0.09, 0.06],
    [0.15, 0.09, 0.06],
    [0.20, 0.12, 0.08]
])

# Calculate the marginal probabilities for rows (pi.) and columns (p.j)
marginal_rows = observed_probabilities.sum(axis=1)
marginal_columns = observed_probabilities.sum(axis=0)

# Calculate the expected probabilities under the null hypothesis of independence (pi. * p.j)
expected_probabilities = np.outer(marginal_rows, marginal_columns)

# Create a table to compare observed and expected probabilities
comparison_table = np.dstack((observed_probabilities, expected_probabilities))

comparison_table, np.isclose(observed_probabilities, expected_probabilities)


(array([[[0.15, 0.15],
         [0.09, 0.09],
         [0.06, 0.06]],
 
        [[0.15, 0.15],
         [0.09, 0.09],
         [0.06, 0.06]],
 
        [[0.2 , 0.2 ],
         [0.12, 0.12],
         [0.08, 0.08]]]),
 array([[ True,  True,  True],
        [ True,  True,  True],
        [ True,  True,  True]]))

# **Part 2:**

In [5]:
import numpy as np

# Function to classify each pseudo-random number into one of the nine cells
def classify_number(x, cumulative_probabilities):
    # Find the index of the first cumulative probability that is greater than the random number
    cell = np.where(cumulative_probabilities > x)[0][0]
    return cell

# Observed probabilities from the 3x3 table (flattened for easier cumulative probability calculation)
observed_probabilities_flat = np.array([0.15, 0.09, 0.06, 0.15, 0.09, 0.06, 0.20, 0.12, 0.08])

# Generate the cumulative probability thresholds for the flattened probabilities
cumulative_probabilities = np.cumsum(observed_probabilities_flat)

# Number of observations to generate
num_observations = 300

# Generate 300 pseudo-random numbers between 0 and 1
random_numbers = np.random.uniform(0, 1, num_observations)

# Initialize an array to hold the counts for each cell
cell_counts_corrected = np.zeros(9, dtype=int)

# Classify each pseudo-random number into the nine cells based on the cumulative probabilities
for x in random_numbers:
    cell = classify_number(x, cumulative_probabilities)
    cell_counts_corrected[cell] += 1

# Reshape the counts to the 3x3 table format
sample_observations_corrected = cell_counts_corrected.reshape(3, 3)
sample_observations_corrected


array([[52, 29, 17],
       [43, 23, 20],
       [54, 43, 19]])

# **Part 3:**
in note.

# **Part 4:**

In [6]:
import numpy as np
import scipy.stats as stats

# Function to simulate observed Q values and perform the goodness-of-fit test
def test_chi_squared_goodness_of_fit(observed_values, df, num_bins):
    # Calculate the chi-squared distribution with df degrees of freedom
    chi2_distribution = stats.chi2(df=df)

    # Use the percent point function to find the chi2 values that correspond to the desired percentiles
    bin_edges = chi2_distribution.ppf(np.linspace(0, 1, num_bins + 1))

    # The expected frequency for each bin, assuming the chi-squared distribution
    expected_probs = chi2_distribution.cdf(bin_edges[1:]) - chi2_distribution.cdf(bin_edges[:-1])
    expected_frequencies = expected_probs * len(observed_values)

    # Binning the observed Q values according to the defined bins
    observed_frequencies, _ = np.histogram(observed_values, bins=bin_edges)

    # Perform the chi-squared goodness-of-fit test
    chi2_statistic, p_value = stats.chisquare(observed_frequencies, f_exp=expected_frequencies)

    return chi2_statistic, p_value, observed_frequencies, expected_frequencies, bin_edges

# Generate a sample of Q values using a chi-squared distribution with 4 degrees of freedom for demonstration
np.random.seed(0)  # Seed for reproducibility
simulated_observed_Q_values = np.random.chisquare(4, size=30)

# Perform the test with the simulated observed Q values
chi2_statistic, p_value, observed_freq, expected_freq, bin_edges = test_chi_squared_goodness_of_fit(
    simulated_observed_Q_values, df=4, num_bins=5
)

chi2_statistic, p_value, observed_freq, expected_freq, bin_edges


(5.66666666666667,
 0.22546314129598105,
 array([9, 2, 4, 7, 8]),
 array([6., 6., 6., 6., 6.]),
 array([0.        , 1.64877662, 2.75284268, 4.04462649, 5.98861669,
               inf]))

The results from the execution of this code are as follows:

Chi-Squared Statistic:
χ2≈5.67
p-value:
≈
p≈0.225
Observed Frequencies: [9, 2, 4, 7, 8]
Expected Frequencies: [6.0, 6.0, 6.0, 6.0, 6.0]
Bin Edges: [0.0, 1.65, 2.75, 4.04, 5.99, inf]